internal/controller/vrg_kubeobjects.go

```diff
@@ -210,52 +216,108 @@ func (v *VRGInstance) kubeObjectsCaptureStartOrResume(
+
+	annotations := map[string]string{vrgGenerationKey: strconv.FormatInt(generation, vrgGenerationNumberBase)}
+	labels := util.OwnerLabels(v.instance)
+
+	allEssentialStepsFailed, err := v.executeCaptureSteps(result, pathName, capturePathName, namePrefix,
+		veleroNamespaceName, captureInProgressStatusUpdate, annotations, requests, log)
+	if err != nil {
+		return
+	}
+
+	if allEssentialStepsFailed {
+		v.kubeObjectsCaptureFailed("KubeObjectsCaptureError", "Kube objects capture failed")
+
+		return
+	}
+
+	firstRequest := getFirstRequest(captureSteps, requests, namePrefix, v.s3StoreAccessors[0].S3ProfileName)
+	v.kubeObjectsCaptureComplete(
+		result,
+		captureStartConditionally,
+		captureNumber,
+		veleroNamespaceName,
+		interval,
+		labels,
+		firstRequest.StartTime(),
+		firstRequest.Object().GetAnnotations(),
+	)
+}
+
+//nolint:gocognit,funlen,cyclop
+func (v *VRGInstance) executeCaptureSteps(result *ctrl.Result, pathName, capturePathName, namePrefix,
+	veleroNamespaceName string, captureInProgressStatusUpdate captureInProgressStatusUpdate,
+	annotations map[string]string, requests map[string]kubeobjects.Request, log logr.Logger,
+) (bool, error) {
+	captureSteps := v.recipeElements.CaptureWorkflow
+	failOn := v.recipeElements.CaptureFailOn
+	allEssentialStepsFailed := true
+	essentialStepsCount := 0
 	requestsProcessedCount := 0
 	requestsCompletedCount := 0
-	annotations := map[string]string{vrgGenerationKey: strconv.FormatInt(generation, vrgGenerationNumberBase)}
+	labels := util.OwnerLabels(v.instance)
+
+	for groupNumber, captureGroup := range captureSteps {
+		var err error
+
+		var loopCount int
 
-	for groupNumber, captureGroup := range groups {
 		cg := captureGroup
 		log1 := log.WithValues("group", groupNumber, "name", cg.Name)
+		isEssentialStep := cg.GroupEssential != nil && *cg.GroupEssential
 
 		if cg.IsHook {
-			if err := v.executeHook(cg.Hook, log1); err != nil {
-				break
-			}
-		} else {
-			requestsCompletedCount += v.kubeObjectsGroupCapture(
+			err = v.executeHook(cg.Hook, log1)
+		}
+
+		if !cg.IsHook {
+			loopCount, err = v.kubeObjectsGroupCapture(
 				result, cg, pathName, capturePathName, namePrefix, veleroNamespaceName,
 				captureInProgressStatusUpdate,
 				labels, annotations, requests, log,
 			)
+			requestsCompletedCount += loopCount
+		}
 
-			requestsProcessedCount += len(v.s3StoreAccessors)
-			if requestsCompletedCount < requestsProcessedCount {
-				log1.Info("Kube objects group capturing", "complete", requestsCompletedCount, "total", requestsProcessedCount)
+		if err != nil {
+			if shouldStopExecution(failOn, isEssentialStep) {
+				v.kubeObjectsCaptureFailed("KubeObjectsWorkflowError", err.Error())
 
-				return
+				return false, err
 			}
+
+			allEssentialStepsFailed = allEssentialStepsFailed && isEssentialStep
+
+			continue
+		}
+
+		if isEssentialStep {
+			// shows that at least one essential step has succeeded
+			allEssentialStepsFailed = false
```

Comment:
This is disabled for all `captureSteps`. In other words, if `isEssentialSteps` is always `true`, line 295 seems redundant. I'm not sure what the code is trying to determine.